A54-restrict-control-plane-status-codes.md

```diff
@@ -0,0 +1,136 @@
+could be implemented as a gRPC call to a service. That service might fail the
+RPC with `NOT_FOUND` because the host is not registered. But if the data plane
+has an RPC reading a file `funny-jokes.txt`, the RPC should fail with
+`UNAVAILABLE`. Failing the RPC with `NOT_FOUND` would confuse the application
+into believing `funny-jokes.txt` doesn't exist. This is the situation that
+caused a serious [Spotify outage][] because of [a bug in gRPC Java][issue 8950].
+
+[statuscodes.md]: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+[Spotify outage]: https://engineering.atspotify.com/2022/03/incident-report-spotify-outage-on-march-8/
+[issue 8950]: https://github.com/grpc/grpc-java/issues/8950
+
+## Proposal
+
+Add checks to each language's Channel that prevents the control plane from
+passing "inappropriate statuses" to the data plane. Specifically:
+
+ * When applying a failed Service Config parsing result from the name resolver
+ * When applying a failed Config Selector result from the name resolver (an
+   internal API for xDS)
+ * When applying a failed Picker result from the load balancer
+
+The check locations are precise and apply only to the places where the control
+plane passes status codes to the data plane. At the time of this gRFC, this list
+is exhaustive. Future gRFCs should add similar requirements in the case they add
+an API that passes status codes from the control plane to the data plane.
+
+"Inappropriate statuses" are statuses whose code is one of:
+
+ * `OK` (in some APIs this is okay, but it must not be copied to an RPC)
+ * `INVALID_ARGUMENT`
+ * `NOT_FOUND`,
+ * `ALREADY_EXISTS`
+ * `FAILED_PRECONDITION`
+ * `ABORTED`
+ * `OUT_OF_RANGE`
+ * `DATA_LOSS`
+
+If an "inappropriate status" is detected, the status should be rewritten to use
+`INTERNAL` status code and include text within the status message making it
+clear a bug occurred. The original status code and message should be included
+within the new status message so the error information is not lost.
+
+Retry policy status code matching may only be checked after the control plane
+status check and rewrite. However, if an inappropriate status is rewritten and
+the retry policy allows retrying `INTERNAL`, implementations are free to not
+retry the RPC.
+
+### Temporary environment variable protection
+
+No temporary environment variable protection will be used.
+
+## Rationale
+
+This gRFC does not make it "okay" for a name resolver or load balancer to
+propagate status codes verbatum from the control plane. Such behavior would
+still be considered a bug and the `INTERNAL` error code will cause such bugs to
+fail noticably. The point of the gRFC is to reduce the damage to the system when
+these bugs are triggered. This does not improve reliability directly, but it
+does ease recovery.
+
+This gRFC purposefully does not limit interceptors, even interceptors that
+communicate with a control plane (e.g., RLQS). An interceptor like fault
```

Comment:
Most readers probably don't know what "RLQS" is, and we don't yet have a gRFC describing it.  How about instead just saying something like "an interceptor communicating with an external control plane for rate-limiting purposes"?

Reply:
That's a lot of words and specificity for a theoretical interceptor. I think I'd sooner remove the RQLS reference and provide no example. I thought that even if you didn't know RQLS it was helpful to those that do know it. I'm now linking to the .proto.

Fault injection now links to the gRFC as well.